src/translate.py

```diff
@@ -2583,38 +2599,48 @@ def array_access_from_add(
         pass
     else:
         # base->subarray[index]
-        substr_array = find_substruct_array(base.type, offset, scale)
-        if substr_array is None:
+        sub_path, sub_type, remaining_offset = base.type.get_deref_field(
+            offset, target_size=scale
+        )
+        # Check if the last item in the path is `0`, which indicates the start of an array
+        # If it is, remove it: it will be replaced by `[index]`
+        if sub_path is None or len(sub_path) < 2 or sub_path[-1] != 0:
             return None
-        sub_field_name, sub_offset, elem_type = substr_array
+        sub_path.pop()
         base = StructAccess(
             struct_var=base,
-            offset=sub_offset,
+            offset=offset - remaining_offset,
             target_size=None,
-            field_name=sub_field_name,
+            field_path=sub_path,
             stack_info=stack_info,
-            type=Type.ptr(elem_type),
+            type=sub_type,
         )
-        offset -= sub_offset
-        target_type = elem_type
+        offset = remaining_offset
+        target_type = sub_type
 
-    # Add .field if necessary
     ret: Expression = ArrayAccess(base, index, type=target_type)
-    field_name, new_type, ptr_type, array_dim = get_field(
-        base.type, offset, target_size=target_size
+
+    # Add .field if necessary by wrapping ret in StructAccess(AddressOf(...))
+    ret_ref = AddressOf(ret, type=ret.type.reference())
+    field_path, field_type, remaining_offset = ret_ref.type.get_deref_field(
+        offset, target_size=target_size
     )
+    if remaining_offset != 0:
+        field_path = None
+
     if offset != 0 or (target_size is not None and target_size != scale):
         ret = StructAccess(
-            struct_var=AddressOf(ret, type=Type.ptr()),
+            struct_var=ret_ref,
             offset=offset,
             target_size=target_size,
-            field_name=field_name,
+            field_path=field_path,
             stack_info=stack_info,
-            type=ptr_type if array_dim is not None else new_type,
+            type=field_type,
         )
 
-    if ptr and array_dim is None:
-        ret = AddressOf(ret, type=ptr_type)
+    if ptr:
```

Comment:
This is an interesting diff -- it shows that we previously *did* avoid AddressOf wrapping for arrays like I was thinking in my earlier comment. I feel like the difference should be visible if the result is dereferenced? Admittedly the type juggling logic in this function loses me completely, so it may for all I know be equivalent or more correct. We could push looking into this to future work.

A disadvantage of hiding `&` in formatting of AddressOf-with-array-type is that it covers up bugs with this pretty well!

Reply:
Yeah - it's a good catch.
The previous version of the code seems tenuous to me -- if the inner field is an array (`array_dim is not None`), then the returned expression doesn't depend on `ptr`? 

I think the new version is more consistent -- though at the cost of decaying arrays.